File: WayforkWindows/app/lib/core/links/subscription_fetcher.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:wayfork/core/links/proxy_link_parser.dart';
import 'package:wayfork/core/version.dart';

/// Fetches a subscription body for `SubscriptionDecoder`
/// (docs/design/04-tunnels.md, "Subscriptions"): https only, redirects
/// followed, 15 s, 1 MiB, a neutral User-Agent so converters answer with raw
/// links rather than Clash YAML. The URL is a bearer token — callers must not
/// log or store it; this class never does.
abstract final class SubscriptionFetcher {
  static const timeout = Duration(seconds: 15);
  static const maxBodyBytes = 1048576;

  static Future<String> fetch(Uri url, {HttpClient? client}) async {
    if (url.scheme.toLowerCase() != 'https') {
      throw const ProxyLinkException(
        ProxyLinkError.unsupported,
        'subscriptions must use https',
      );
    }
    final http = client ?? HttpClient();
    http.connectionTimeout = timeout;
    http.userAgent = 'Wayfork/${WayforkVersion.app}';
    try {
      final HttpClientResponse response;
      final List<int> bytes;
      try {
        final request = await http.getUrl(url).timeout(timeout);
        request.headers.set(HttpHeaders.acceptHeader, 'text/plain, */*;q=0.1');
        response = await request.close().timeout(timeout);
        bytes = await response
            .fold<List<int>>([], (all, chunk) => all..addAll(chunk))
            .timeout(timeout);
      } on Object catch (error) {
        throw ProxyLinkException(ProxyLinkError.invalid, _describe(error));
      }
      final finalUrl = response.redirects.isEmpty
          ? url
          : response.redirects.last.location;
      if (finalUrl.scheme.toLowerCase() != 'https' && finalUrl.hasScheme) {
        throw const ProxyLinkException(
          ProxyLinkError.unsupported,
          'subscription redirected away from https',
        );
      }
      if (response.statusCode < 200 || response.statusCode > 299) {
        throw ProxyLinkException(
          ProxyLinkError.invalid,
          'server answered ${response.statusCode}',
        );
      }
      if (bytes.length > maxBodyBytes) {
        throw const ProxyLinkException(
          ProxyLinkError.invalid,
          'subscription is larger than 1 MiB',
        );
      }
      try {
        return utf8.decode(bytes);
      } on FormatException {
        throw const ProxyLinkException(
          ProxyLinkError.invalid,
          'subscription is not text',
        );
      }
    } finally {
      if (client == null) http.close(force: true);
    }
  }

  static String _describe(Object error) => switch (error) {
    TimeoutException() => 'timed out',
    SocketException(:final message) => message,
    HttpException(:final message) => message,
    _ => '$error',
  };
}

File: macos/WayforkCore/Sources/WayforkCore/Links/SubscriptionFetcher.swift
import Foundation

/// Fetches a subscription body for `SubscriptionDecoder` (docs/design/04-tunnels.md,
/// "Subscriptions"): https only, redirects followed, 15 s, 1 MiB, a neutral User-Agent so
/// converters answer with raw links rather than Clash YAML. The URL is a bearer token —
/// callers must not log or store it; this type never does.
public enum SubscriptionFetcher {
    public static let timeout: TimeInterval = 15
    public static let maxBodyBytes = 1_048_576

    public static func fetch(_ url: URL, session: URLSession = .shared) async throws -> String {
        guard url.scheme?.lowercased() == "https" else {
            throw ProxyLinkError.unsupported("subscriptions must use https")
        }
        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.setValue("Wayfork/\(WayforkCore.version)", forHTTPHeaderField: "User-Agent")
        request.setValue("text/plain, */*;q=0.1", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw ProxyLinkError.invalid(error.localizedDescription)
        }
        guard let http = response as? HTTPURLResponse else {
            throw ProxyLinkError.invalid("no HTTP response")
        }
        guard http.url?.scheme?.lowercased() == "https" else {
            throw ProxyLinkError.unsupported("subscription redirected away from https")
        }
        guard (200...299).contains(http.statusCode) else {
            throw ProxyLinkError.invalid("server answered \(http.statusCode)")
        }
        guard data.count <= maxBodyBytes else {
            throw ProxyLinkError.invalid("subscription is larger than 1 MiB")
        }
        guard let body = String(data: data, encoding: .utf8) else {
            throw ProxyLinkError.invalid("subscription is not text")
        }
        return body
    }
}
